fiftyone/core/collections.py

```diff
@@ -10548,42 +10523,132 @@ def create_index(self, field_or_spec, unique=False, wait=True, **kwargs):
+            _existing_unique = True
+
+        # Handle existing indexes
+        convert_to_unique = False
+        replace_existing = False
+        if _existing_name is not None:
+            if db_name != _existing_db_name and force:
+                # Replace existing index with different orderings
+                replace_existing = True
+            elif unique and not _existing_unique:
+                # Upgrade existing index to unique
+                convert_to_unique = True
+            elif _existing_unique and not unique and force:
+                # Downgrade existing index to non-unique
+                replace_existing = True
+            elif _existing_unique or (unique == _existing_unique):
+                # Satisfactory index already exists
+                return _existing_name
+
+            # Handle default indexes
+            if (
+                _index_name in self._get_default_indexes(frames=is_frame_index)
+                and index_name != "filepath"  # allow 'filepath' to be modified
+            ):
+                raise ValueError(f"Cannot modify default index '{index_name}'")
+
+        # Convert existing index to unique, if necessary
+        if convert_to_unique:
+            logger.info(
+                f"Converting existing index '{index_name}' to unique "
+                f"on dataset '{self._dataset.name}'"
             )
 
-        name = coll.create_index(index_spec, unique=unique, **kwargs)
+            db = foo.get_db_conn()
+
+            try:
+                # https://www.mongodb.com/docs/manual/core/index-unique/convert-to-unique
+                db.command(
+                    "collMod",
+                    coll_name,
+                    index={"name": _existing_db_name, "prepareUnique": True},
+                )
+                db.command(
+                    "collMod",
+                    coll_name,
+                    index={"name": _existing_db_name, "unique": True},
+                )
+
+                return _existing_name
+            except:
+                if foo.get_db_version() < Version("6"):
+                    # index conversion was introduced in MongoDB 6
+                    replace_existing = True
+                else:
+                    db.command(
+                        "collMod",
+                        coll_name,
+                        index={
+                            "name": _existing_db_name,
+                            "prepareUnique": False,
+                        },
+                    )
+
+                    raise
+
+        # Drop existing index, if necessary
+        if replace_existing:
+            if not convert_to_unique:
+                logger.info(
+                    f"Replacing existing index '{index_name}' "
+                    f"on dataset '{self._dataset.name}'"
+                )
+
+            coll.drop_index(_existing_db_name)
+
+        # Create new index, if necessary
+        if not wait:
```

Comment:
I know this isn't really changed by this pr, but setting w=0 makes the index operation return immediately, but subsequent operations in the same python session seem to hang. For example 

```
import fiftyone as fo
ds=fo.load_dataset('big-image-ds')
ds.create_index('db_pos') # returns immediately
ds = fo.load_dataset('quickstart') # hangs forever
```

```
import fiftyone as fo
ds=fo.load_dataset('big-image-ds')
ds.create_index('index'')  #returns immediately
ds._sample_collection.find_one({}) #hangs forever
```

Reply:
Not sure as the index build would have taken awhile as well for that dataset. 

Created ticket in the backlog to investigate https://voxel51.atlassian.net/browse/FOEPD-2223